File: subprojects/basic03/basic-tutorial-3.py
#!/usr/bin/env python3

"""
Basic Tutorial 3: Dynamic Pipelines
"""

import sys
import logging

import gi
gi.require_version("Gst", "1.0")
gi.require_version("GLib", "2.0")
from gi.repository import GLib, Gst

logging.basicConfig(level=logging.INFO, format="[%(levelname)8s] - %(message)s")
logger = logging.getLogger(__name__)

class CustomData:
    """CustomData contains all the information needed to pass to callbacks
    """
    def __init__(self, pipeline, source, convert, resample, sink):
        self.pipeline = pipeline
        self.source = source
        self.convert = convert
        self.resmaple = resample
        self.sink = sink


def on_pad_added(src, new_pad, data):
    """This function will be called by the pad-added signal
    """
    sink_pad = data.convert.get_static_pad("sink")
    # If our converter is already linked, we have nothing to do here
    if (sink_pad.is_linked()):
        logger.error("We are already linked. Ignoring.")
        return

    # Retrieves the current capabilities of the new pad
    new_pad_caps = new_pad.get_current_caps()
    logger.info(f"Received new pad '{new_pad_caps[0].get_name()}' from '{src.get_name()}'")

    # Check the new pad's type
    new_pad_struct = new_pad_caps.get_structure(0)
    new_pad_type = new_pad_struct.get_name()
    if not new_pad_type.startswith("audio/x-raw"):
        logger.info(f"It has type '{new_pad_type}' which is not raw audio. Ignoring.")
        return

    try:
        new_pad.link(sink_pad)
    except:
        logger.error(f"Type is '{new_pad_type}' but link failed.")
    else:
        logger.info(f"Link succeeded (type '{new_pad_type}')")


def on_bus_message(bus, message, loop):
    if message.type == Gst.MessageType.EOS:
        logger.info("End-of-stream reached.")
        loop.quit()
    elif message.type == Gst.MessageType.ERROR:
        err, debug_info = message.parse_error()
        logger.error(f"Error received from element {message.src.get_name()}: {err.message}")
        logger.error(f"Debugging information: {debug_info if debug_info else 'none'}")
        loop.quit()
    elif message.type == Gst.MessageType.STATE_CHANGED:
        # We are only interested in state-changed messages from the pipeline
        old_state, new_state, pending_state = message.parse_state_changed()
        if ("test-pipeline" == message.src.get_name()):
            logger.info(f"Pipeline state changed: {old_state.value_nick} -> {new_state.value_nick}")
    else:
        logger.debug(f"Unhandled message type {message.type}.")
    return True


def main(argv):
    # Initialize GStreamer
    Gst.init(None)

    uri = "https://gstreamer.freedesktop.org/data/media/sintel_trailer-480p.webm"
    if len(argv) != 2:
        logger.info(f"usage: {argv[0]} <uri>")
        logger.info(f"missing uri argument, use default uri instead.")
    elif Gst.uri_is_valid(argv[1]):
        uri = argv[1]

    # Create the elements
    source = Gst.ElementFactory.make("uridecodebin", "source")
    convert = Gst.ElementFactory.make("audioconvert", "convert")
    resample = Gst.ElementFactory.make("audioresample", "resample")
    sink = Gst.ElementFactory.make("autoaudiosink", "sink")

    # Create the empty pipeline
    pipeline = Gst.Pipeline.new("test-pipeline")

    # Build the pipeline.
    pipeline.add(source, convert, resample, sink)
    # Note that we are NOT linking the source at this point. We will do it later
    Gst.Element.link_many(convert, resample, sink)
    # Set the URI to play
    source.set_property("uri", uri)
    # Prepare data for callback
    pdata = CustomData(pipeline, source, convert, resample, sink)
    # Connect to the pad-added signal
    source.connect("pad-added", on_pad_added, pdata)

    loop = GLib.MainLoop()

    # Listen to the bus
    bus = pipeline.get_bus()
    bus.add_signal_watch()
    bus.connect("message", on_bus_message, loop)

    # start play back and listen to events
    ret = pipeline.set_state(Gst.State.PLAYING)
    if ret == Gst.StateChangeReturn.FAILURE:
        logger.error("Unable to set the pipeline to the playing state.")
        return -1

    try:
        loop.run()
    except:
        logger.error("Unhandled exception.")

    # cleanup
    pipeline.set_state(Gst.State.NULL)


if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: subprojects/basic03/basic-tutorial-3.cpp
/* gstreamermm - a C++ wrapper for gstreamer
 *
 * Basic Tutorial 3: Dynamic Pipelines
 */

#include <gstreamermm.h>
#include <glibmm/main.h>
#include <glibmm/stringutils.h>
#include <iostream>
#include <cstdlib>

Glib::RefPtr<Glib::MainLoop> mainloop;

// This function is used to receive asynchronous messages in the main loop.
bool on_bus_message(const Glib::RefPtr<Gst::Bus>& /* bus */,
    const Glib::RefPtr<Gst::Message>& message)
{
  switch (message->get_message_type()) {
    case Gst::MESSAGE_EOS:
      std::cout << std::endl << "End of stream" << std::endl;
      mainloop->quit();
      return false;
    case Gst::MESSAGE_ERROR:
    {
      Glib::RefPtr<Gst::MessageError> msgError {Glib::RefPtr<Gst::MessageError>::cast_static(message)};
      if (msgError)
      {
        Glib::Error err {msgError->parse_error()};
        std::string debug_info {msgError->parse_debug()};
        std::cerr << "Error received from element " << message->get_source()->get_name() << ": " <<
            err.what() << std::endl;
        if (!debug_info.empty())
        {
          std::cout << "Debugging information: " << debug_info << std::endl;
        }
      }
      else
      {
        std::cerr << "Error." << std::endl;
      }
      mainloop->quit();
      return false;
    }
    case Gst::MESSAGE_STATE_CHANGED:
    {
      // We are only interested in state-changed messages from the pipeline
      if ("test-pipeline" == message->get_source()->get_name())
      {
        auto state_get_name = [] (Gst::State state) -> std::string {
          return gst_element_state_get_name(static_cast<GstState>(state));
        };
        Glib::RefPtr<Gst::MessageStateChanged> msgSC {Glib::RefPtr<Gst::MessageStateChanged>::cast_static(message)};
        Gst::State old_state {msgSC->parse_old_state()};
        Gst::State new_state {msgSC->parse_new_state()};
        std::cout << "Pipeline state changed: " <<
            state_get_name(old_state) << " -> " <<
            state_get_name(new_state) << std::endl;
      }
      break;
    }
    default:
        //std::cout << "Unhandled message type: " << message->get_message_type() << std::endl;
      break;
  }

  return true;
}

int main(int argc, char** argv)
{
  // Initialize gstreamermm:
  Gst::init(argc, argv);

  // default uri
  Glib::ustring uri {"https://gstreamer.freedesktop.org/data/media/sintel_trailer-480p.webm"};

  // Take the commandline argument and ensure that it is a uri:
  if (argc < 2)
  {
    std::cout << "Usage: " << argv[0] << " <uri>" << std::endl;
    std::cout << "missing uri argument, use default uri instead." << std::endl;
  }
  else if (Gst::URIHandler::uri_is_valid(argv[1]))
  {
    uri = argv[1];
  }

  // Create elements
  Glib::RefPtr<Gst::Element> source {Gst::ElementFactory::create_element("uridecodebin", "source")},
    convert {Gst::ElementFactory::create_element("audioconvert", "convert")},
    resample {Gst::ElementFactory::create_element("audioresample", "resample")},
    sink {Gst::ElementFactory::create_element("autoaudiosink", "sink")};

  if (!source || !convert || !resample || !sink)
  {
    std::cerr << "One of the elements could not be created." << std::endl;
    return EXIT_FAILURE;
  }

  // Create the empty pipeline
  Glib::RefPtr<Gst::Pipeline> pipeline {Gst::Pipeline::create("test-pipeline")};

  // add the elements to the pipeline before linking them
  try
  {
    pipeline->add(source)->add(convert)->add(resample)->add(sink);
  }
  catch (std::runtime_error& ex)
  {
    std::cerr << "Exception while adding: " << ex.what() << std::endl;
    return EXIT_FAILURE;
  }

  // Link the elements
  try
  {
    // We link the elements converter, resample and sink, but we DO NOT link them with the source,
    // since at this point it contains no source pads. We do it later in a pad-added signal handler.
    convert->link(resample)->link(sink);
  }
  catch(const std::runtime_error& ex)
  {
    std::cout << "Exception while linking elements: " << ex.what() << std::endl;
  }

  // Set the uri property.
  source->set_property("uri", uri);
  // Signal handler for on-pad-added signal of source element
  // Here we use lambda to expose local variables that are needed
  source->signal_pad_added().connect(
    [convert] (const Glib::RefPtr<Gst::Pad> &new_pad)
    {
      Glib::RefPtr<Gst::Pad> sink_pad {convert->get_static_pad("sink")};
      // If our converter is already linked, we have nothing to do here
      if (sink_pad->is_linked())
      {
        std::cout << "sink pad of convert is already linked. Ignoring." << std::endl;
        return;
      }
      // Retrieves the current capabilities of the new pad
      Glib::RefPtr<Gst::Caps> new_pad_caps {new_pad->get_current_caps()}; 
      Glib::ustring media_type {new_pad_caps->get_structure(0).get_name()};
      std::cout << "Received new pad " << new_pad->get_name() << ", media type: " << media_type << std::endl;
      // Check the new pad's type
      if (Glib::str_has_prefix(media_type, "audio/x-raw"))
      {
        Gst::PadLinkReturn ret = new_pad->link(sink_pad);
        if (ret != Gst::PAD_LINK_OK && ret != Gst::PAD_LINK_WAS_LINKED)
        {
          std::cerr << "Linking of pads " << new_pad->get_name() << " and " <<
              sink_pad->get_name() << " failed." << std::endl;
        }
      }
      else
      {
        std::cout << "Media type is not raw audio. Ignoring." << std::endl;
      }
    });

  // Create the main loop.
  mainloop = Glib::MainLoop::create();

  // Get the bus and watch the messages
  Glib::RefPtr<Gst::Bus> bus {pipeline->get_bus()};
  bus->add_watch(sigc::ptr_fun(&on_bus_message));

  // start play back and listen to events
  if (pipeline->set_state(Gst::STATE_PLAYING) == Gst::STATE_CHANGE_FAILURE)
  {
    std::cerr << "Unable to set the pipeline to the playing state." << std::endl;
    return EXIT_FAILURE;
  }

  // Now set the playbin to the PLAYING state and start the main loop:
  std::cout << "Running." << std::endl;
  mainloop->run();

  // Clean up nicely:
  std::cout << "Returned. Stopping pipeline." << std::endl;
  pipeline->set_state(Gst::STATE_NULL);

  return EXIT_SUCCESS;
}
